lib/Monads/trace/Trace_RG.thy

```diff
@@ -230,6 +314,135 @@ lemma bind_twp[wp_split]:
   apply (clarsimp simp: guar_cond_append)
   done
 
+lemma bindE_twp[wp_split]:
+  "\<lbrakk>\<And>r. \<lbrace>Q' r\<rbrace>,\<lbrace>R\<rbrace> g r \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>; \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q'\<rbrace>,\<lbrace>E\<rbrace>\<rbrakk>
+   \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>=E (\<lambda>r. g r) \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  apply (simp add: bindE_def validIE_def)
+  apply (erule bind_twp[rotated])
+  apply (clarsimp simp: lift_def throwError_def split: sum.splits)
+  apply (subst validI_valid_no_trace_eq)
+   apply wpsimp+
+  done
+
+\<comment> \<open>FIXME MC: improve naming for bind_twp vs seq_ext vs rg_seq_ext
+          existing proofs mention hoare_seq_ext a lot and seq_ext a few times\<close>
+\<comment> \<open>FIXME MC: names of seq_ext and seq_extE clash with valid equivalents\<close>
```

Comment:
I'd recommend `(* .. *)` comments for FIXME, because then you're not constrained by LaTeX syntax (these here will crash on the underscore, I think).

Reply:
The `seq_ext` names are old and a bit non-standard. If we can eliminate them now in favour of variations of `bind_wp`, I'd be up for it.

Reply:
> I'd recommend `(* .. *)` comments for FIXME, because then you're not constrained by LaTeX syntax (these here will crash on the underscore, I think).

Right. I have my Jedit set up to default to `\<comment> \<open> .. \<close>`, since it works inside lemma statements, but `(* .. *)` would clearly be better here. Quite a few of the FIXMEs will be removed after I incorporate all of your feedback and I'll update the remaining ones afterwards. If nothing else, that should get the tests green.

Reply:
> The `seq_ext` names are old and a bit non-standard. If we can eliminate them now in favour of variations of `bind_wp`, I'd be up for it.

I'll come up with something that makes sense to me and see what you think. That would be another thing that would be good to bring back to `nondet`.

Reply:
I'm currently going with renaming `seq_ext` to `bind_twp_fwd` and removing `rg_seq_ext` completely. See the most recent commit for the additional changes it leads to.

What do you think? If you're happy with that then I'll add it to the list of changes to make to `Nondet_VCG` and `Trace_VCG`.